source/simulator/helper_functions.cc

```diff
@@ -909,8 +909,164 @@ namespace aspect
+    const double min_advection_exact_global = (advection_field.is_temperature()
+                                               ?
+                                               parameters.global_temperature_min_preset
+                                               :
+                                               parameters.global_composition_min_preset[advection_field.compositional_variable]
+                                              );
+
+
+    LinearAlgebra::BlockVector distributed_vector (introspection.index_sets.system_partitioning,
+                                                   mpi_communicator);
+    unsigned int block_idx = advection_field.block_index(introspection);
+    distributed_vector.block(block_idx) = solution.block(block_idx);
+
+    std::vector<types::global_dof_index> local_dof_indices (finite_element.dofs_per_cell);
+
+    typename DoFHandler<dim>::active_cell_iterator
+    cell = dof_handler.begin_active(),
+    endc = dof_handler.end();
+    for (; cell != endc; ++cell)
+      {
+        if (cell->is_locally_owned())
+          {
+            cell->get_dof_indices (local_dof_indices);
+            fe_values_12.reinit (cell);
+            fe_values_21.reinit (cell);
+
+            fe_values_12[field].get_function_values(solution, values_12);
+            fe_values_21[field].get_function_values(solution, values_21);
+
+            //Find the local max and local min
+            double min_advection_local =  std::numeric_limits<double>::max();
+            double max_advection_local = -std::numeric_limits<double>::max();
+            for (unsigned int q = 0; q < n_q_points_12; ++q)
+              {
+                min_advection_local = std::min<double> (min_advection_local, values_12[q]);
+                max_advection_local = std::max<double> (max_advection_local, values_12[q]);
+              }
+            for (unsigned int q = 0; q < n_q_points_21; ++q)
+              {
+                min_advection_local = std::min<double> (min_advection_local, values_21[q]);
+                max_advection_local = std::max<double> (max_advection_local, values_21[q]);
+              }
+
+            //Find the trouble cell
+            if (   min_advection_local < min_advection_exact_global
+                   || max_advection_local > max_advection_exact_global )
+              {
+                //Compute the cell area and cell advection average
+                double local_area = 0;
+                for (unsigned int q = 0; q < n_q_points_12; ++q)
+                  {
+                    local_area += fe_values_12.JxW(q);
+                  }
+
+
+                double local_advection_average = 0;
+                for (unsigned int q = 0; q < n_q_points_12; ++q)
+                  {
+                    local_advection_average += values_12[q]*fe_values_12.JxW(q);
+                  }
+                local_advection_average /= local_area;
+                //Define theta: which is a scaling constant
```

Comment:
Hmm, this comment does not make the meaning of theta_T clearer to me. What about: `Define theta, a scaling constant that limits the local deviation from the local advection field average to the maximum possible value considering the user-given exact bounds` ? Also: is the `_T` important? Does it stand for temperature? Because then we can remove it (it is used for the composition as well now).